### ISSUE TITLE
Add 30 sec. countdown on the Browser Assistant icon (during `Do not filter for 30 seconds`)

### ISSUE THREAD

**Issue by contribucious** [AdguardTeam/BrowserAssistant#80]
#### Hi, :wave:
&nbsp;
It might be a good idea IMHO to display the 30 second countdown on the extension icon, when using the `Don't filter for 30 seconds` option.

Useful, as we don't know where we are in time when we close the Browser Assistant popup and use the website. :relaxed:

:small_blue_diamond: **Possible subtlety:** trigger its display only when the popup is closed. This, in order to prevent a double display of the countdown while the popup is still open — i.e. on the extension icon + in the popup. And … leave it or not on the extension icon if/when the popup is reopened (up to you).
&nbsp;

## Visual rendering

![AdGuard_f6FhzH4Zha](https://user-images.githubusercontent.com/4764956/184269094-eba4eb85-3d93-4491-a4e4-1a63d854d76d.png)
###### :arrow_right_hook: _Like this but with a gray shield and displaying `19s` instead — if there are `19s` left in the countdown for example. :hourglass_flowing_sand:_

:information_source: **Versus blocking counter visually?** In addition to the presence of the character "s", and the decrease every second clearly recognizable, all this in addition on a _gray_ shield in this situation and as a result of the user's _immediate own action_, to be possibly even more visually distinct from a possible/probable future blocking counter — already present in `AdGuard Browser extension` —, just change the background color + text color + possibly _round_ the corners or display it in a round or oval instead. And/Or … either display an _animated-or-not_ hourglass (:hourglass_flowing_sand:) or a vertical bar (such as a capital i) that decreases from top to bottom, at his side.

:bulb: **Future-proof:** Once the `30s` delay is over, the shield will turn green again (once [this issue](https://github.com/AdguardTeam/BrowserAssistant/issues/79) is fixed though) and … a potential future blocking counter will take over instead. To count the new blocked XHR & co requests for example. Not incompatible with the future, therefore. :heavy_check_mark:

&nbsp;
___
#### _Thank you in advance. :thumbsup:_
_Environment_
**Browser Assistant** 1.3.13 on **Firefox** 103.0.2
**AdGuard** 7.10.2 on **Windows** 10 Pro 21H2 (OS Build 19044.1865)

**Comment by contribucious** [comment]
### :mag: Possible visual result

#### Main

<details>
<summary>View it …</summary>

![w2BLYXPt7yMiPVKT2_HLvQ](https://user-images.githubusercontent.com/4764956/184335236-688aa38d-2297-4614-8f95-9b3390d3a4f2.gif)
([Source image](https://miro.medium.com/max/1200/1*w2BLYXPt7yMiPVKT2_HLvQ.gif) / [Source webpage](https://medium.com/swlh/unity-tips-simple-ui-countdown-81a16074c5d4)) — Ideal one (colors to be adapted however). 3 characters in it.
</details>

#### Alternative

<details>
<summary>View it …</summary>

![D6e1c9A5z3](https://user-images.githubusercontent.com/4764956/184335883-4b81c773-3cab-487c-9983-ac4f280752a0.jpg)
([Source image](https://www.jqueryscript.net/images/Circular-Countdown-Clock-Plugin-For-jQuery-CircularCountDownJs.jpg) / [Source webpage](https://www.jqueryscript.net/time-clock/Circular-Countdown-Clock-Plugin-For-jQuery-CircularCountDownJs.html)) — Simplified version, if there is not enough space for 3 characters. Pure graphic version (i.e. like this image but … _without_ any character inside). So, more [like so](https://user-images.githubusercontent.com/4764956/184336803-7a889d73-19c7-41ac-9ffd-b0c3bb404a6a.jpg) in fact ([source image](https://i.pinimg.com/originals/79/a1/34/79a1345034619d70145a13590aa23f96.jpg) / [source webpage](https://www.pinterest.com/pin/395331673518544248/)).
</details>

#### Another alternative _(variant, less visible in small size as a badge however)_

<details>
<summary>View it …</summary>

![rJgzoWGbdWAsBCU4](https://user-images.githubusercontent.com/4764956/184337273-3d834b40-dbe8-432d-ae15-e9b45ad0f7d9.png)
([Source image](https://miro.medium.com/max/1838/0*rJgzoWGbdWAsBCU4.png) / [Source webpage](https://medium.com/@DrawandCode/how-to-draw-an-animated-timer-in-android-b4951304881e)) — Variant. Nothing inside either.
</details>

**Comment by Birbber** [comment]
@contribucious Great idea, will consider implementing it.

**Comment by maximtop** [comment]
We can't style the badge as you proposed in the second comment.

**Comment by contribucious** [comment]
@maximtop

You're right. [Badge](https://developer.chrome.com/docs/extensions/reference/browserAction/#badge) is very limited in fact — unlike [Icon](https://developer.chrome.com/docs/extensions/reference/browserAction/#icon) — regarding animation.

**Badge** (`chrome.browserAction.setBadgeText`) seems OK for a countdown (30 … 29 … 28 …), but it's the **Icon** (`chrome.browserAction.setIcon`) that should be updated for a possible animation.

##### :gear: Although possible [using canvas](https://timleland.com/animated-badge-icon/) _(note: despite the generic "badge" word in the blog post title, _technically_ it's the icon which is animated there)_ ([source](https://stackoverflow.com/a/47066852)), this in order "[to create more dynamic UIs](https://developer.chrome.com/docs/extensions/reference/browserAction/#icon)", it's more typically used for a "loading" (i.e. perpetual spinning) icon visibly.

###### :arrow_right_hook: Need to see if possible regarding a possible "slow" countdown one, more adapted. Even possibly a very basic one like a "I" with 3 vertical bricks (like a "AA battery"): from 30 to 20 (3/3 shown), from 20 to 10 (2/3 shown), from 10 to 0 (1/3 shown).

_But unfortunately, if I understand correctly, during the 30 seconds, the gray shield will have to be replaced by this animation (which uses canvas). We cannot have both at the same time, except … to create a very complex canvas._

**Comment by maximtop** [comment]
> to create a very complex canvas.
I prefer it to be an easy solution

**Comment by contribucious** [comment]
Of course. I just mentioned this as a bonus, as the only possibility AFAIK to do this (i.e. to have both animation + keeping the gray shield icon), but not as a real de facto solution. :wink:

**Comment by zubrRB** [comment]
Keep in mind that there are plans to add blocked element statistics to the icon.

**Comment by contribucious** [comment]
@maximtop

I think the simplest and most adequate solution here will be _just_ to update the badge itself (30s … 29s … 28s …) using `chrome.browserAction.setBadgeText`.

And possibly to use [`chrome.browserAction.setBadgeBackgroundColor`](https://developer.chrome.com/docs/extensions/reference/browserAction/#badge) with a more pronounced color during these 30 seconds to differentiate`*` from a future probable blocking counter to come (as already said in my issue, @zubrRB :wink:).
&nbsp;
___
`*` In addition to the "s" character (i.e. `30s` vs `30`) + the decreasing each second clearly recognizable + the "gray shield only" situation + the fact that this follows an immediate user action (i.e. clicking on `Do not filter for 30 seconds`). All already described in **`Versus blocking counter visually?`** and **`Future-proof`** parts of my issue.

**Comment by gork7777** [comment]
@contribucious @maximtop I edited the badge little. This contour will start at 12 o'clock and decrease clockwise. I think it could be better this way.

<details><summary>Badge</summary>

![edit](https://user-images.githubusercontent.com/35523759/184715819-0a214f82-b41b-4dc5-9958-1aeb33dc82e8.png)
</details>

**Comment by contribucious** [comment]
@gork7777 Very nice one! :thumbsup: Although the icon should be gray during these 30 seconds (as it is now), since the protection is switched off during this time (action is `Do not filter for 30 seconds`). :wink:

@maximtop To avoid having to use canvas, especially because not required here (i.e. no "perpetual loading" icon with complex animation), could we imagine having simply 30x the static image of @gork7777 for each second of the countdown (i.e. `icon-30s.png`, `icon-29s.png`, …) and just updating this static image every second using `chrome.browserAction.setIcon` in an appropriate `setInterval`?

_(And possibly, as a bonus to this already explicit visual, displaying the number of seconds like `29s` using `chrome.browserAction.setBadgeText` and `chrome.browserAction.setBadgeBackgroundColor`?)_

**Comment by contribucious** [comment]
## Update
&nbsp;
If this feature sees the light of day, it is better to go for [its light version](https://github.com/AdguardTeam/BrowserAssistant/issues/80#issuecomment-1214865930).

Because unfortunately, here is what the **Manifest V3** [documentation](https://developer.chrome.com/docs/extensions/reference/action/#icon) says:
> The `action.setIcon()` API is intended to set a static image. It should not be used to simulate animation.

###### _And, just for info/context ([source](https://developer.chrome.com/docs/extensions/reference/action/#emulating-pageactions-with-declarativecontent)):_
> The chrome.action API replaced the [browserAction](https://developer.chrome.com/docs/extensions/reference/browserAction/) and [pageAction](https://developer.chrome.com/docs/extensions/reference/pageAction/) APIs in Manifest V3.
